File: scripts/pathform.py
#!/usr/bin/env python3
"""pathform - accept a file path in EITHER form and return one that opens.

WHY THIS EXISTS. Bash on this box takes /c/Users/..., Windows-native python takes
C:/Users/.... Handing the wrong form across the boundary raises FileNotFoundError on a
file that demonstrably exists and was read seconds earlier by the other language.

THREE INSTANCES IN NINETY MINUTES, TWO AGENTS, NEITHER NEW TO THIS BOX: seb_boss on a
/tmp procs.csv, builder_1 at 00:06Z (the root cause behind a 3,991-character task
description being destroyed), seb_boss again while VERIFYING that fix. Nobody was
careless in any of the three, which is the whole argument: a rule saying "remember which
form python wants" IS THE THING THAT ALREADY FAILED THREE TIMES. This makes the error
impossible rather than infrequent.

SHARPEST CONSEQUENCE, AND THE REASON IT IS NOT COSMETIC: A VERIFICATION TOOL FAILING
INTO THE SHAPE OF THE THING IT VERIFIES. The third instance was a retained-input check
reporting a missing file, which is indistinguishable from the input actually being lost.
The instrument built to prove nothing was lost can manufacture the exact evidence that
something was.

    from pathform import resolve
    open(resolve(p), encoding="utf-8")

    python pathform.py --self-test
"""
import os
import re
import sys

__all__ = ["forms", "resolve"]


def forms(p):
    """Every plausible spelling of `p`, original first. Pure string work, no I/O."""
    p = str(p)
    out = [p]
    # MSYS -> Windows:  /c/Users/x  ->  C:/Users/x
    m = re.match(r"^/([A-Za-z])/(.*)$", p)
    if m:
        out.append(f"{m.group(1).upper()}:/{m.group(2)}")
    # Windows -> MSYS:  C:\Users\x or C:/Users/x  ->  /c/Users/x
    m = re.match(r"^([A-Za-z]):[\\/](.*)$", p)
    if m:
        out.append("/" + m.group(1).lower() + "/" + m.group(2).replace("\\", "/"))
    # Backslashes alone, no drive letter.
    if "\\" in p:
        out.append(p.replace("\\", "/"))
    seen, uniq = set(), []
    for c in out:
        if c not in seen:
            seen.add(c)
            uniq.append(c)
    return uniq


def resolve(p, must_exist=True):
    """Return the first spelling of `p` that exists.

    RAISES ON A GENUINELY ABSENT PATH, and that is not a detail. A helper that only ever
    succeeds would have fixed the false negative by making a real missing file INVISIBLE
    — strictly worse than the bug it replaces, and it would pass every positive case
    above. The error names every form tried, because a raise that hides what it looked
    for is as uninformative as the FileNotFoundError it replaces.
    """
    tried = forms(p)
    for cand in tried:
        if os.path.exists(cand):
            return cand
    if not must_exist:
        return tried[0]
    raise FileNotFoundError(
        f"no form of {p!r} exists; tried: " + ", ".join(repr(t) for t in tried))


def _self_test():
    import tempfile
    fd, win = tempfile.mkstemp(suffix="-pathform.txt")
    os.write(fd, b"ok")
    os.close(fd)
    win = win.replace("\\", "/")
    m = re.match(r"^([A-Za-z]):/(.*)$", win)
    assert m, win
    msys = "/" + m.group(1).lower() + "/" + m.group(2)
    failed = 0
    try:
        cases = [
            # The two forms this language is handed in practice. THE MSYS ONE IS THE
            # CASE THAT FAILS TODAY — a test using only the Windows form proves nothing,
            # which is precisely how this survived three occurrences.
            ("windows form opens", lambda: open(resolve(win)).read() == "ok"),
            ("MSYS form opens", lambda: open(resolve(msys)).read() == "ok"),
            ("backslash form opens", lambda: open(resolve(win.replace("/", "\\"))).read() == "ok"),
            ("both forms resolve to the SAME file", lambda: os.path.samefile(resolve(win), resolve(msys))),
            # NEGATIVE. Without this the helper could return its input unchanged and pass
            # every case above while making a real missing file invisible.
            ("a genuinely absent path still RAISES", _absent_raises),
            ("the raise NAMES the forms it tried", _raise_names_forms),
            ("must_exist=False returns rather than raising", lambda: resolve("/c/no/such", must_exist=False) == "/c/no/such"),
        ]
        for name, fn in cases:
            try:
                ok = fn() is True
            except Exception:
                ok = False
            failed += 0 if ok else 1
            print(("ok   " if ok else "FAIL ") + name)
    finally:
        os.unlink(win)
    print()
    print(f"pathform --self-test: {len(cases) - failed}/{len(cases)}")
    print("BOUNDARY: proves both forms open FROM PYTHON. The node twin (pathform.mjs) "
          "covers the other language; the four-combination claim needs BOTH run.")
    return failed == 0


def _absent_raises():
    try:
        resolve("/c/Users/definitely-not-here-4a7f/x.txt")
    except FileNotFoundError:
        return True
    return False


def _raise_names_forms():
    try:
        resolve("/c/Users/definitely-not-here-4a7f/x.txt")
    except FileNotFoundError as e:
        return "C:/Users/definitely-not-here-4a7f/x.txt" in str(e)
    return False


if __name__ == "__main__":
    if "--self-test" in sys.argv:
        sys.exit(0 if _self_test() else 2)
    if len(sys.argv) > 1:
        print(resolve(sys.argv[1]))
    else:
        print(__doc__.strip().splitlines()[0])

File: scripts/pathform.mjs
// pathform - accept a file path in EITHER form and return one that opens. Node twin of
// scripts/pathform.py; see that file for the full incident record.
//
// Bash on this box takes /c/Users/..., node on Windows takes C:/Users/.... Node resolves
// a bare /tmp to C:\tmp, which is NOT where MSYS bash wrote it. THREE INSTANCES IN
// NINETY MINUTES, TWO AGENTS. The 00:06Z one cost 3,991 characters of a task
// description: node ENOENT'd on a heredoc file, `cat` printed nothing, and the empty
// string reached a CLI that accepted it as content.
//
//   import { resolve } from './pathform.mjs';
//   readFileSync(resolve(p), 'utf8');
//
//   node scripts/pathform.mjs --self-test
import { existsSync, statSync } from 'node:fs';

export function forms(p) {
  p = String(p);
  const out = [p];
  const msys = /^\/([A-Za-z])\/(.*)$/.exec(p);
  if (msys) out.push(`${msys[1].toUpperCase()}:/${msys[2]}`);
  const win = /^([A-Za-z]):[\\/](.*)$/.exec(p);
  if (win) out.push(`/${win[1].toLowerCase()}/${win[2].replace(/\\/g, '/')}`);
  if (p.includes('\\')) out.push(p.replace(/\\/g, '/'));
  return [...new Set(out)];
}

// RAISES ON A GENUINELY ABSENT PATH — the negative case is the point. A helper that only
// ever succeeds fixes the false negative by making a real missing file INVISIBLE, which
// is strictly worse than the bug, and it passes every positive test. The message names
// every form tried, because a throw that hides what it looked for is no better than the
// ENOENT it replaces.
export function resolve(p, { mustExist = true } = {}) {
  const tried = forms(p);
  for (const c of tried) if (existsSync(c)) return c;
  if (!mustExist) return tried[0];
  throw new Error(`no form of ${JSON.stringify(p)} exists; tried: ${tried.join(', ')}`);
}

if (process.argv.includes('--self-test')) {
  const { mkdtempSync, writeFileSync, rmSync } = await import('node:fs');
  const { join } = await import('node:path');
  const { tmpdir } = await import('node:os');
  const dir = mkdtempSync(join(tmpdir(), 'pathform-'));
  const win = join(dir, 'probe.txt').replace(/\\/g, '/');
  writeFileSync(win, 'ok', 'utf8');
  const m = /^([A-Za-z]):\/(.*)$/.exec(win);
  const msys = `/${m[1].toLowerCase()}/${m[2]}`;
  const { readFileSync } = await import('node:fs');
  const cases = [
    // THE MSYS CASE IS THE ONE THAT FAILS TODAY. A test using only the Windows form
    // proves nothing — that is exactly how this survived three occurrences.
    ['windows form opens', () => readFileSync(resolve(win), 'utf8') === 'ok'],
    ['MSYS form opens', () => readFileSync(resolve(msys), 'utf8') === 'ok'],
    ['backslash form opens', () => readFileSync(resolve(win.replace(/\//g, '\\')), 'utf8') === 'ok'],
    ['both forms resolve to the SAME inode', () =>
      statSync(resolve(win)).ino === statSync(resolve(msys)).ino],
    // NEGATIVE. Without it the helper could return its input unchanged, pass everything
    // above, and turn a real missing file into a silent success.
    ['a genuinely absent path still THROWS', () => {
      try { resolve('/c/Users/definitely-not-here-4a7f/x.txt'); return false; }
      catch { return true; }
    }],
    ['the throw NAMES the forms it tried', () => {
      try { resolve('/c/Users/definitely-not-here-4a7f/x.txt'); return false; }
      catch (e) { return e.message.includes('C:/Users/definitely-not-here-4a7f/x.txt'); }
    }],
    ['mustExist:false returns rather than throwing', () =>
      resolve('/c/no/such', { mustExist: false }) === '/c/no/such'],
  ];
  let failed = 0;
  for (const [name, fn] of cases) {
    let ok = false;
    try { ok = fn() === true; } catch { ok = false; }
    if (!ok) failed += 1;
    console.log((ok ? 'ok   ' : 'FAIL ') + name);
  }
  rmSync(dir, { recursive: true, force: true });
  console.log('');
  console.log(`pathform.mjs --self-test: ${cases.length - failed}/${cases.length}`);
  console.log('BOUNDARY: proves both forms open FROM NODE. The python twin covers the');
  console.log('other language; the four-combination claim needs BOTH run.');
  process.exit(failed === 0 ? 0 : 2);
}
